v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/ConvertChangeStream.java

```diff
@@ -0,0 +1,252 @@
+      }
+
+      c.output(KV.of(hbaseMutations.getRow(), hbaseMutations));
+    }
+
+    /**
+     * Checks if mutation was replicated from HBase.
+     *
+     * @param mutation from change stream
+     * @return true if mutation was replicated from hbase
+     */
+    private boolean isHbaseReplicated(ChangeStreamMutation mutation, String hbaseQualifierInput) {
+      List<Entry> mutationEntries = mutation.getEntries();
+
+      if (mutationEntries.size() == 0) {
+        return false;
+      }
+      Entry lastEntry = mutationEntries.get(mutationEntries.size() - 1);
+
+      if (lastEntry instanceof DeleteCells) {
+        if (((DeleteCells) lastEntry)
+            .getQualifier()
+            .equals(ByteString.copyFromUtf8(hbaseQualifierInput))) {
+          LOG.info(
+              "Filtering: {}",
+              mutation.toRowMutation("").toProto(RequestContext.create("", "", "")));
+          Metrics.counter("HbaseRepl", "mutations_filtered_from_hbase").inc();
+          return true;
+        }
+      }
+      Metrics.counter("HbaseRepl", "mutations_valid_from_bigtable").inc();
+      return false;
+    }
+
+    static byte[] convertUtf8String(String utf8) {
+      return utf8.getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Appends origin information to row mutation.
+     *
+     * @param hbaseMutations row mutation to append origin info to
+     * @param cbtQualifierInput origin info string denoting mutation is from bigtable
+     * @throws IOException
+     */
+    private void appendOriginInfoToMutations(RowMutations hbaseMutations, String cbtQualifierInput)
+        throws IOException {
+      // Get last mutation to deduce last column family from
+      Mutation lastMutation =
+          hbaseMutations.getMutations().get(hbaseMutations.getMutations().size() - 1);
+      List<String> familyList = (List<String>) lastMutation.getFingerprint().get("families");
+      String lastEntryCf = familyList.get(familyList.size() - 1);
+      // If this is null it means we didn't process any mutation entries. this
+      // shouldn't be possible once all mutation types are supported
+      if (lastEntryCf != null) {
+        Delete hiddenDelete = new Delete(hbaseMutations.getRow(), 0L);
+        hiddenDelete.addColumns(
+            convertUtf8String(lastEntryCf), convertUtf8String(cbtQualifierInput));
+        hbaseMutations.add(hiddenDelete);
+      }
+    }
+
+    /**
+     * Converts Cdc ChangeStreamMutation to HBase RowMutations object.
+     *
+     * @param mutation changeStreamMutation
+     * @return Hbase RowMutations object
+     */
+    private static RowMutations convertToRowMutations(ChangeStreamMutation mutation)
+        throws IOException {
+      byte[] hbaseRowKey = mutation.getRowKey().toByteArray();
+      RowMutations hbaseMutations = new RowMutations(hbaseRowKey);
+
+      for (Entry entry : mutation.getEntries()) {
+        if (entry instanceof SetCell) {
+
```

Comment:
Can we create a separate convert method for each Mutation entry type? Since there's a decent amount of logic for some of these and it'll be easier to unit test

Reply:
added convert functions for setCells, deleteCells, and deleteFamily